contrib/zmq/zmq_sub.py

```diff
@@ -1,44 +1,65 @@
-#!/usr/bin/env python2
+#!/usr/bin/env python3
 # Copyright (c) 2014-2016 The Bitcoin Core developers
 # Distributed under the MIT software license, see the accompanying
 # file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+# A blocking example using python 2.7 can be obtained from the git history:
+# https://github.com/bitcoin/bitcoin/blob/37a7fe9e440b83e2364d5498931253937abe9294/contrib/zmq/zmq_sub.py
+
 import array
 import binascii
-import zmq
+import asyncio, zmq, zmq.asyncio
+import signal
 import struct
 
 port = 28332
 
-zmqContext = zmq.Context()
-zmqSubSocket = zmqContext.socket(zmq.SUB)
-zmqSubSocket.setsockopt(zmq.SUBSCRIBE, "hashblock")
-zmqSubSocket.setsockopt(zmq.SUBSCRIBE, "hashtx")
-zmqSubSocket.setsockopt(zmq.SUBSCRIBE, "rawblock")
-zmqSubSocket.setsockopt(zmq.SUBSCRIBE, "rawtx")
-zmqSubSocket.connect("tcp://127.0.0.1:%i" % port)
-
-try:
-    while True:
-        msg = zmqSubSocket.recv_multipart()
-        topic = str(msg[0])
+class ZMQHandler():
+    def __init__(self):
+        self.loop = zmq.asyncio.install()
+        self.zmqContext = zmq.asyncio.Context()
+
+        self.zmqSubSocket = self.zmqContext.socket(zmq.SUB)
+        self.zmqSubSocket.setsockopt_string(zmq.SUBSCRIBE, "hashblock")
+        self.zmqSubSocket.setsockopt_string(zmq.SUBSCRIBE, "hashtx")
+        self.zmqSubSocket.setsockopt_string(zmq.SUBSCRIBE, "rawblock")
+        self.zmqSubSocket.setsockopt_string(zmq.SUBSCRIBE, "rawtx")
+        self.zmqSubSocket.connect("tcp://127.0.0.1:%i" % port)
+
+    async def handle(self) :
```

Comment:
Asyncio was only introduced in 3.5 including these syntax enhancements and was "provisional".  It's now final that 3.6 was released a week ago, and backported (as final) to 3.5 point releases.  3.4 doesn't have it and never will.  But, it doesn't have anywhere near the user base of 2.7 and I don't think is worth supporting long term.

I tried to install 3.4 while testing this, it's not even provided anymore by Ubuntu.

http://stackoverflow.com/questions/30191556/coroutine-in-python-between-3-4-and-3-5-how-can-i-keep-backwords-compatibility